crates/common/src/l2.rs

```diff
@@ -413,27 +485,34 @@ where
     /// 4. Maintains metrics about syncing progress
     #[instrument(name = "L2Syncer", skip_all)]
     pub async fn run(mut self, mut shutdown_signal: GracefulShutdown) {
-        let (l2_tx, mut l2_rx) = mpsc::channel(1);
-        let l2_sync_worker = sync_l2(
+        tokio::spawn(sync_l2(
             self.start_l2_height,
             self.sequencer_client.clone(),
-            l2_tx,
+            self.block_queue.clone(),
             self.sync_blocks_count,
-        );
-        tokio::pin!(l2_sync_worker);
+        ));
+
+        if let Some(ws_endpoint) = self.sequencer_ws_endpoint.clone() {
+            tokio::spawn(run_subscription_task(ws_endpoint, self.block_queue.clone()));
+        }
 
         let backup_manager = self.backup_manager.clone();
+
+        let notifier = self.block_queue.lock().await.notifier();
         loop {
             select! {
                 biased;
                 _ = &mut shutdown_signal => {
                     info!("Shutting down L2 syncer");
-                    l2_rx.close();
                     return;
                 },
-                Some(l2_blocks) = l2_rx.recv() => {
-                    // While syncing, we'd like to process L2 blocks as they come without any delays.
-                    for l2_block in l2_blocks {
+                _ = notifier.notified() => {
+                    let blocks_to_process = {
+                        let mut queue = self.block_queue.lock().await;
+                        queue.drain_sequential()
+                    };
+
+                    for l2_block in blocks_to_process {
```

Comment:
I believe we have to rethink this approach.

Notice that we drain all the blocks, for loop through them and for us to be stuck at processing the current l2_block, we `loop` indefinitely inside.

Instead of all this, we use the queue to get first unprocessed block, try processing it, if it fails, we don't pop it or drain it out of the queue, it just stays there. This would introduce yet another problem which is the queue getting larger as the operating sequencer gets more blocks. So if the current block fails to process 3 times, we should fail and stop the node.